v2/spiffetls/listen.go

```diff
@@ -24,20 +26,78 @@ func Listen(ctx context.Context, network, laddr string, authorizer tlsconfig.Aut
 //
 // This function is the same as NewListenerWithMode using the MTLSServer mode.
 func NewListener(ctx context.Context, inner net.Listener, authorizer tlsconfig.Authorizer, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+	return NewListenerWithMode(ctx, inner, MTLSServer(authorizer), options...)
 }
 
 // ListenWithMode creates a TLS listener accepting connections on the given
 // network address using net.Listen. The given mode controls the authentication
 // and authorization exercised during the TLS handshake.
 func ListenWithMode(ctx context.Context, network, laddr string, mode ListenMode, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+	inner, err := net.Listen(network, laddr)
+	if err != nil {
+		return nil, err
+	}
+
+	l, err := NewListenerWithMode(ctx, inner, mode, options...)
+	if err != nil {
+		inner.Close()
+		return nil, err
+	}
+
+	return l, nil
 }
 
 // NewListenerWithMode creates a TLS listener which accepts connections from an
 // inner Listener and wraps each connection with tls.Server. The given mode
 // controls the authentication and authorization exercised during the TLS
 // handshake.
-func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (_ net.Listener, err error) {
+	m := mode.get()
+	source := m.source
+	if source == nil {
+		source, err = workloadapi.NewX509Source(ctx, m.options...)
+		// Close source if there is a failiure after this point
+		defer func() {
+			if err != nil {
+				source.Close()
+			}
+		}()
+		if err != nil {
+			return nil, spiffetlsErr.New("cannot create X.509 source: %w", err)
+		}
+	}
+
+	if m.bundle == nil {
+		m.bundle = source
+	}
+	if m.svid == nil {
+		m.svid = source
+	}
+
+	opt := &listenConfig{}
+	for _, option := range options {
+		option.apply(opt)
+	}
+
+	tlsConfig := &tls.Config{}
+	if opt.baseTLSConf != nil {
+		tlsConfig = opt.baseTLSConf
+	}
+
+	switch m.tlsType {
+	case typeTLSServer:
+		tlsconfig.HookTLSServerConfig(tlsConfig, m.svid)
+	case typeMTLSServer:
+		tlsconfig.HookMTLSServerConfig(tlsConfig, m.svid, m.bundle, m.authorizer)
+	case typeMTLSWebServer:
+		tlsCert, err := tlsconfig.GetTLSCertificate(m.svid)
```

Comment:
The web server does not serve an X509-SVID and should not get the certificate from the source. It serves the certificate passed into the various `MTLSWebServer*` modes, i.e. `m.cert`.